### PR TITLE
ADBDEV-4320: Fixed deadlock if the cross slice SharedScan consumer is on slice 0.

### PR DESCRIPTION
**Pull request by KnightMurloc**
Fix deadlock if the cross slice SharedScan consumer is on slice 0.

When the consumer of the cross slice SharedScan is in slice 0 and the producer
is in another slice, a deadlock may occur. The deadlock could happen because QD
does not complete slice 0 until other slices will be completed. But the
producer's slice cannot finish until it receives notification from the consumer
about the completion of reading. The consumer does not send this notification
before all QEs are completed.

This patch fixes the deadlock by creating a list of cross-slice consumers for
the current slice and sending notifications before waiting for QEs completion.
We can omit the share_lk_ctxt check for NULL since only cross slice Shared Scans
are added to the list.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/55195

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/657573

**Comment by RekGRpth** [comment]
Can the second commit work without the first?

**Comment by KnightMurloc** [comment]
Technically, the test case can work without the prefetch fix, since the executor goes to the inner for its cost. but I'm not sure about the work in general.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/55922

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/55923

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/679140

File: src/include/nodes/execnodes.h [hunk 1]
```diff
@@ -661,6 +661,9 @@ typedef struct EState
 
 	/* Should the executor skip past the alien plan nodes */
 	bool eliminateAliens;
+
+	/* list of Sharedscan consumers in the current slice */
+	List		*sharedScanConsumers;
```
Comment (RekGRpth) [inline src/include/nodes/execnodes.h]:
```suggestion
	List	   *sharedScanConsumers;
```

File: src/backend/executor/execUtils.c [hunk 2]
```diff
@@ -2098,6 +2099,24 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
...
 
+	/*
+	 * If the reading part of SharedScan is in slice 0, then we will not notify
+	 * the writing part that we have finished reading before get the query result.
+	 * However, we will not get the result until the writing part shutdown.
+	 * The writing part will wait for a notify from the reader to shutdown.
+	 * To prevent this we send a notify to writer before we receive result.
+	 */
+	ListCell *cell;
+	foreach(cell, estate->sharedScanConsumers)
```
Comment (RekGRpth) [inline src/backend/executor/execUtils.c]:
```suggestion
	foreach (cell, estate->sharedScanConsumers)
```

File: src/backend/executor/execUtils.c [hunk 3]
```diff
@@ -2098,6 +2099,24 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
...
 	}
 
+	/*
+	 * If the reading part of SharedScan is in slice 0, then we will not notify
+	 * the writing part that we have finished reading before get the query result.
+	 * However, we will not get the result until the writing part shutdown.
+	 * The writing part will wait for a notify from the reader to shutdown.
+	 * To prevent this we send a notify to writer before we receive result.
+	 */
+	ListCell *cell;
```
Comment (RekGRpth) [inline src/backend/executor/execUtils.c]:
```suggestion
	ListCell   *cell;
```

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/56471

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/700017

**Comment by RekGRpth** [comment]
Let's look at the following two plans without patch. Why, when executing, doesn't the first one freeze, but the second one freezes?
```sql
SET gp_cte_sharing = on;
SET optimizer = off;
CREATE TABLE d (c1 int, c2 int) DISTRIBUTED BY (c1);
CREATE TABLE r (c1 int, c2 int) DISTRIBUTED replicated;
INSERT INTO d (VALUES ( 2, 0 ),( 2 , 0 ));
INSERT INTO r (VALUES ( 2, 0 ),( 2 , 0 ));
EXPLAIN (COSTS OFF, analyze off, verbose off) WITH cte AS (
	SELECT c1 FROM d LIMIT 2
) SELECT * FROM cte a JOIN (SELECT * FROM r JOIN cte USING (c1) LIMIT 1) b USING (c1);
                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 Hash Join
   Hash Cond: (share0_ref2.c1 = b.c1)
   ->  Shared Scan (share slice:id 0:0)
   ->  Hash
         ->  Subquery Scan on b
               ->  Limit
                     ->  Hash Join
                           Hash Cond: (share0_ref1.c1 = r.c1)
                           ->  Shared Scan (share slice:id 0:0)
                                 ->  Materialize
                                       ->  Limit
                                             ->  Gather Motion 3:1  (slice1; segments: 3)
                                                   ->  Limit
                                                         ->  Seq Scan on d
                           ->  Hash
                                 ->  Gather Motion 1:1  (slice2; segments: 1)
                                       ->  Seq Scan on r
 Optimizer: Postgres query optimizer
(18 rows)
EXPLAIN (COSTS OFF, analyze off, verbose off) WITH cte AS (
	SELECT c1 FROM r LIMIT 2
) SELECT * FROM cte a JOIN (SELECT * FROM d JOIN cte USING (c1) LIMIT 1) b USING (c1);
                                       QUERY PLAN                                       
----------------------------------------------------------------------------------------
 Hash Join
   Hash Cond: (share0_ref2.c1 = b.c1)
   ->  Shared Scan (share slice:id 0:0)
   ->  Hash
         ->  Subquery Scan on b
               ->  Limit
                     ->  Gather Motion 3:1  (slice2; segments: 3)
                           ->  Limit
                                 ->  Hash Join
                                       Hash Cond: (d.c1 = share0_ref1.c1)
                                       ->  Seq Scan on d
                                       ->  Hash
                                             ->  Redistribute Motion 1:3  (slice1)
                                                   Hash Key: share0_ref1.c1
                                                   ->  Shared Scan (share slice:id 1:0)
                                                         ->  Materialize
                                                               ->  Limit
                                                                     ->  Seq Scan on r
 Optimizer: Postgres query optimizer
(19 rows)
```

**Comment by KnightMurloc** [comment]
> Why, when executing, doesn't the first one freeze, but the second one freezes?

in the first case, SharedScan occurs inside a single slice, and in this case, no read completion messages are sent, and therefore deadlock does not occur. I will correct the description that the problem is related only to cross slice Shared Scan.

**Comment by RekGRpth** [comment]
> the problem is related only to cross slice Shared Scan

Maybe we need to add some kind of condition for checking cross-slices?

**Comment by KnightMurloc** [comment]
there is a `share_lk_ctxt` structure that is created only in the case of cross slicer and the current patch checks this. the only thing is that we can move the check to the stage of filling the list.
https://github.com/arenadata/gpdb/blob/215d8d8f9fec39300e4eb84977ade4fce040ebc1/src/backend/executor/nodeShareInputScan.c#L237-L241

File: src/backend/executor/execUtils.c [hunk 4]
```diff
@@ -2098,6 +2099,24 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
...
+	 * the writing part that we have finished reading before get the query result.
+	 * However, we will not get the result until the writing part shutdown.
+	 * The writing part will wait for a notify from the reader to shutdown.
+	 * To prevent this we send a notify to writer before we receive result.
+	 */
+	ListCell   *cell;
+	foreach (cell, estate->sharedScanConsumers)
+	{
+		ShareInputScanState *state = lfirst(cell);
+		if (state->share_lk_ctxt)
```
Comment (RekGRpth) [inline src/backend/executor/execUtils.c]: Is this check redundant?

Comment (KnightMurloc) [reply src/backend/executor/execUtils.c]: in the current patch, no, because if SharedScan occurs inside one slide, it will be NULL.

Comment (KnightMurloc) [reply src/backend/executor/execUtils.c]: ok, I checked it looks like for non-cross slice Shared Scan driver_slice is -1 and such SharedScan is not added to the list so we can remove this condition.

Comment (RekGRpth) [reply src/backend/executor/execUtils.c]: and shareinput_reader_waitready expects non-null first argument ;)

**Comment by RekGRpth** [comment]
Maybe a reading shared scan in a zero slice on segments?

**Comment by KnightMurloc** [comment]
the slice 0 is always on the coordinator as I know.

**Comment by RekGRpth** [comment]
> the slice 0 is always on the coordinator as I know.

Not always, insert for example

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/56566

**Comment by KnightMurloc** [comment]
I managed to get only SharedScan in one slice for insert. but technically there should be no problems since this patch changes the logic not only for the coordinator.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/56677

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/708921

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/708917

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/708918

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/56723

**Comment by BenderArenadata** [comment]
Failed job Publish allure report link: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/710489

**Review by RekGRpth — DISMISSED** [review]

File: src/test/regress/sql/with.sql [hunk 5]
```diff
@@ -1081,5 +1081,17 @@ WITH cte AS (
 	SELECT count(*) c1 FROM d
 ) SELECT * FROM cte a JOIN (SELECT * FROM d JOIN cte USING (c1) LIMIT 1) b USING (c1);
 
+-- Test shared scan with consumer on slice 0
+-- the reading part of shared scan must be in the slice 0, otherwise the test becomes useless
```
Comment (bimboterminator1) [inline src/test/regress/sql/with.sql]: Describe what are you testing in more details

File: src/backend/executor/execUtils.c [hunk 6]
```diff
@@ -2098,6 +2099,21 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
 		ExecSquelchNode(queryDesc->planstate);
 	}
 
+	/*
+	 * If the reading part of SharedScan is in slice 0, then we will not notify
+	 * the writing part that we have finished reading before get the query result.
+	 * However, we will not get the result until the writing part shutdown.
+	 * The writing part will wait for a notify from the reader to shutdown.
+	 * To prevent this we send a notify to writer before we receive result.
```
Comment (bimboterminator1) [inline src/backend/executor/execUtils.c]: Really hard to understand, I can hardly define what the original problem was about. Rewrite it entirely, please.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/57352

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/734755

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/734756

**Review by bimboterminator1 — DISMISSED** [review]

File: src/include/nodes/execnodes.h [hunk 7]
```diff
@@ -661,6 +661,9 @@ typedef struct EState
 
 	/* Should the executor skip past the alien plan nodes */
 	bool eliminateAliens;
+
+	/* list of Sharedscan consumers in the current slice */
+	List	   *sharedScanConsumers;
```
Comment (bimboterminator1) [inline src/include/nodes/execnodes.h:666]: @KnightMurloc, Shouldn't we write in the comments that it's a list of only **cross-slice** SharedScans?

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/57528

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/743286

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/743287

**Review by bimboterminator1 — DISMISSED** [review]

**Review by RekGRpth — DISMISSED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/62777

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1009175

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1009176

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1009169

**Comment by BenderArenadata** [comment]
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013142

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/62887

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013722

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013719

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013724

**Review by andr-sokolov — APPROVED** [review]

**Pull request merged**